### PR TITLE
[Bugfix] Fix CPU split-KV scratchpad sizing

### PR DESCRIPTION
**Pull request by gausah01**
## Purpose

Fix a CPU split-KV hang seen with `google/gemma-4-26B-A4B-it`.

The worker could get stuck in `reduce_splits` waiting for a split completion flag. The root cause was the split-KV scratchpad was sized using the final output element size / smaller fallback q-head count, while the runtime split path writes temporary split outputs as `float` and can process more q-heads per split.

This could under-size the split-KV reduction buffer and allow split output data to overlap nearby reduction state, including split completion flags. As a result, `reduce_splits` could wait forever.

This PR sizes the relevant temporary outputs as `float` and sizes the split-KV reduction scratchpad using the maximum q-head tile the runtime can use.

## Test Plan

Test CPU throughput benchmark for `google/gemma-4-26B-A4B-it`

```bash
LD_PRELOAD=/lib/aarch64-linux-gnu/libtcmalloc_minimal.so.4 \
VLLM_TARGET_DEVICE=cpu \
VLLM_CPU_KVCACHE_SPACE=64 \
VLLM_EXECUTE_MODEL_TIMEOUT_SECONDS=7200 \
vllm bench throughput \
  --model google/gemma-4-26B-A4B-it \
  --dataset-name sonnet \
  --num-prompts 128 \
  --override-generation-config '{"temperature": "0.0", "top_p": "1.0"}'
  --max-num-batched-tokens 8192 \
  --num-warmup 128
```

## Test Result

Before this change, the benchmark hung after prompt rendering and repeatedly printed:

`No available shared memory broadcast block found in 60 seconds.
`

After this change, the same benchmark run completes.

```
Throughput: 0.80 requests/s, 925.77 total tokens/s, 102.86 output tokens/s
Total num prompt tokens:  131072
Total num output tokens:  16384
```

### CODE REVIEW

File: csrc/cpu/cpu_attn_impl.hpp [hunk 1]
```diff
@@ -430,7 +430,7 @@ class AttentionScheduler {
     const int32_t max_num_q_token_per_iter = max_num_q_per_iter / q_head_per_kv;
     const int64_t default_tile_size = calcu_default_tile_size(
         cache_size, input.head_dim, input.elem_size, input.q_buffer_elem_size,
-        input.logits_buffer_elem_size, input.output_buffer_elem_size,
+        input.logits_buffer_elem_size, sizeof(float),
```
Comment (bigPYJ1151) [inline csrc/cpu/cpu_attn_impl.hpp]: ```output_buffer_elem_size=attn_impl::partial_output_buffer_t```, I checked all existing impl classes and all of them are ```float```. So I think this is not the cause.

Comment (murste01) [inline csrc/cpu/cpu_attn_impl.hpp]: Allocating based on `sizeof(float)` will over-allocate when the buffer element size is e.g. BF16. If `input.logits_buffer_elem_size, input.output_buffer_elem_size` is too small we should focus on fixing that rather than hardcoding a larger size.

File: csrc/cpu/cpu_attn_impl.hpp [hunk 2]
```diff
@@ -671,22 +671,30 @@ class AttentionScheduler {
...
           ((sc.get_thread_scratchpad_size() + 63) / 64) * 64;
 
+      // Split-KV reduction can use GQA for one-token split tails even when
+      // the scheduler falls back to MHA. Size for the maximum q heads a
+      // runtime split can process, which is enforced by max_num_q_per_iter.
       sc.update(0, metadata_ptr->reduction_split_num, input.head_dim,
-                q_head_per_kv * split_kv_q_token_num_threshold,
-                input.output_buffer_elem_size);
+                max_num_q_per_iter * split_kv_q_token_num_threshold,
+                sizeof(float));
```
Comment (murste01) [inline csrc/cpu/cpu_attn_impl.hpp]: See above.

File: csrc/cpu/cpu_attn_impl.hpp [hunk 3]
```diff
@@ -671,22 +671,30 @@ class AttentionScheduler {
...
 
     {
-      // when q_tile_size = max_num_q_per_iter, requires max
-      // attention_scratchpad_size
+      // Size for the largest q-head tile that the runtime can use.
+      // The runtime rounds q_head_tile_size and passes it into sc.update(),
+      // so sizing only for max_num_q_per_iter can under-allocate here.
       AttentionScratchPad sc(0, *metadata_ptr, 0x0);
+      const int32_t max_q_head_tile_size = default_tile_size;
+      const bool one_round = max_q_head_tile_size <= max_num_q_per_iter;
```
Comment (bigPYJ1151) [inline csrc/cpu/cpu_attn_impl.hpp]: I think this is the exact fix, good catch.

File: csrc/cpu/cpu_attn_impl.hpp [hunk 4]
```diff
@@ -671,22 +671,30 @@ class AttentionScheduler {
...
       metadata_ptr->attention_scratchpad_size_per_thread =
           ((sc.get_thread_scratchpad_size() + 63) / 64) * 64;
 
+      // Split-KV reduction can use GQA for one-token split tails even when
+      // the scheduler falls back to MHA. Size for the maximum q heads a
+      // runtime split can process, which is enforced by max_num_q_per_iter.
       sc.update(0, metadata_ptr->reduction_split_num, input.head_dim,
-                q_head_per_kv * split_kv_q_token_num_threshold,
-                input.output_buffer_elem_size);
+                max_num_q_per_iter * split_kv_q_token_num_threshold,
```
Comment (bigPYJ1151) [inline csrc/cpu/cpu_attn_impl.hpp]: This is not a cause. Currently the KV split is only enabled for sequences with  ```q_token_num==1```. So allocate partial results buffer as ```q_head_per_kv * split_kv_q_token_num_threshold``` is correct.

Comment (gausah01) [reply csrc/cpu/cpu_attn_impl.hpp]: Thanks, I tested the narrower patch with `q_head_per_kv * split_kv_q_token_num_threshold `, but the Gemma 4 run still hangs. Restoring this sizing change to `max_num_q_per_iter * split_kv_q_token_num_threshold` while keeping `input.output_buffer_elem_size`, makes the run complete. so, this q-head sizing change still seems required for this case.

File: csrc/cpu/cpu_attn_impl.hpp (hunk 1 above)
Comment (gausah01) [reply csrc/cpu/cpu_attn_impl.hpp]: Thanks, agreed. I switched this back to the original `input.output_buffer_elem_size`

File: csrc/cpu/cpu_attn_impl.hpp (hunk 2 above)
Comment (gausah01) [reply csrc/cpu/cpu_attn_impl.hpp]: Done

File: csrc/cpu/cpu_attn_impl.hpp (hunk 1 above)
Comment (gausah01) [reply csrc/cpu/cpu_attn_impl.hpp]: Done, thanks! Now switched it back to `input.output_buffer_elem_size`

File: csrc/cpu/cpu_attn_impl.hpp (hunk 4 above)
Comment (bigPYJ1151) [reply csrc/cpu/cpu_attn_impl.hpp]: Hmm, I think the allocation size is right here, at least for Gemma4. Might have some other bugs need to check.

Comment (gausah01) [reply csrc/cpu/cpu_attn_impl.hpp]: Thanks. Since this version fixes the Gemma 4, do you think it is reasonable to keep this PR scoped to unblocking that case and handle any other issue in a follow-up PR if needed?

Comment (bigPYJ1151) [reply csrc/cpu/cpu_attn_impl.hpp]: But it doesn't match with the runtime offset and will cause wrong results:
https://github.com/vllm-project/vllm/blob/08985351f369d3dd6b80bc54ce143ede268e2846/csrc/cpu/cpu_attn_impl.hpp#L1849-L1863

Comment (gausah01) [reply csrc/cpu/cpu_attn_impl.hpp]: Thanks, I think I understand your point now. Would the right direction be to make the allocation and runtime offsets consistent like this?

- size the reduction scratchpad using max_q_head_tile_size, since allocation needs to cover the largest split width
- use q_head_tile_size as the writer-side stride, since partial_output() writes that many q-heads
- use curr_output_head_num as the reducer-side stride, since reduce_splits() reads that many q-heads

This would fix both the reserved buffer size and the spacing used between split outputs at runtime. Does this look like the right fix direction?

Comment (bigPYJ1151) [reply csrc/cpu/cpu_attn_impl.hpp]: It could output correct results, but I'm not sure whether it can resolve the OOB issue. Seems there is no bug in current KV split buffer allocation sizes and access strides, but still have the OOB issue. There are other bugs cause the OOB, but I have no idea about them for now...

Comment (gausah01) [reply csrc/cpu/cpu_attn_impl.hpp]: I dug a bit deeper with range checks. It now shows the OOB is from a mismatch between scheduler allocation and runtime layout.

For Gemma 4, scheduler allocates the reduction output buffer with `allocated_q_head_tile_size=2`, but runtime uses `stride=8` / `active_heads=8` for `head_dim=512`. So the runtime output region goes past the allocated per-KV-head scratchpad limit.

I’m checking the right scheduler-side value to use so allocation matches the runtime split-KV layout.

**Review by fadara01 — CHANGES_REQUESTED** [review]
I haven't thought a lot about the issue or the fix, but this PR fixes the crash with gemma4, however it exposes the same issue to unsloth/gpt-oss-20b-BF16 (which used to work fine before)

**Comment by gausah01** [comment]
I updated the PR with a more general workitem-based sizing fix.

The previous version fixed the Gemma 4 hang by increasing the split-KV reduction allocation, but it was still a broad sizing change. As @bigPYJ1151 pointed that could hide the real OOB issue, and @fadara01 also found that it regressed gpt-oss.

The new version reserves memory based on the real work that will run, rather than using a fixed larger value. This should be safer for other models because it does not special-case Gemma 4, does not broadly over-allocate, and follows the same GQA/non-GQA decision used by the runtime.

I also updated CPU metadata to use the KV-cache spec KV-head count, so the scheduler matches the actual KV cache layout.

Retested for completeness:

- Gemma 4 benchmark completes: 
```
Throughput: 0.52 requests/s, 260.12 total tokens/s, 131.95 output tokens/s
Total num prompt tokens:  31830
Total num output tokens:  32768
```
- gpt-oss benchmark completes
```
Throughput: 0.82 requests/s, 416.26 total tokens/s, 210.89 output tokens/s
Total num prompt tokens:  31911
Total num output tokens:  32768
```

Basic generation sanity also passes for both Gemma 4 and gpt-oss.

So this updated version should address the original Gemma 4 hang without the broader allocation behavior that caused concerns in the previous patch.

Can i get some reviews on this latest patch please? @bigPYJ1151 @fadara01

File: vllm/v1/attention/backends/cpu_attn.py [hunk 5]
```diff
@@ -137,7 +137,7 @@ def __init__(
         self.vllm_config = vllm_config
 
         parallel_config = vllm_config.parallel_config
-        self.num_kv_heads = vllm_config.model_config.get_num_kv_heads(parallel_config)
+        self.num_kv_heads = kv_cache_spec.num_kv_heads
```
Comment (fadara01) [inline vllm/v1/attention/backends/cpu_attn.py:144]: why is this change needed? won't it break model parallel workflows?

Comment (gausah01) [reply vllm/v1/attention/backends/cpu_attn.py:144]: I changed this to match the actual kv-cache layout. since, this valuye is used later when sizing the split-kv scratchpad. I thought it would be safer to derive it from `kv_cache_spec` rather than model_config. I'm not fully sure about all model parallel cases though!

Comment (fadara01) [reply vllm/v1/attention/backends/cpu_attn.py:144]: I had a quick look, and kv_cache_spec.num_kv_heads is the per-rank number, so this won't be problematic.

Though I don't know why this is related to the problem in question.
If you're changing this, doesn't it make sense to also change: 
```
self.num_heads = vllm_config.model_config.get_num_attention_heads(
            parallel_config
```

Comment (gausah01) [reply vllm/v1/attention/backends/cpu_attn.py:144]: Good point, I removed metadata change now and retested it with C++ scratchpad sizing only and it works fine.

**Comment by bigPYJ1151** [comment]
Hmmm, just checked the hang issue still exist. Seems it it very tricky...

File: vllm/v1/attention/backends/cpu_attn.py (hunk 5 above)
Comment (gausah01) [reply vllm/v1/attention/backends/cpu_attn.py:144]: I rechecked it again and it seems like `num_heads` is coming from `model_config` only and not from `kv_cache_spec` I checked with grep: 
`grep -R "class .*KVCacheSpec\|num_kv_heads\|num_heads" -n vllm | head -100`
I don’t see an equivalent `kv_cache_spec.num_heads`; `KVCacheSpec` only seems to expose `num_kv_heads`. So I think only `num_kv_heads` should come from `kv_cache_spec`, while `num_heads` should remain from `model_config`.

Comment (gausah01) [reply vllm/v1/attention/backends/cpu_attn.py:144]: > Good point, I removed metadata change now and retested it with C++ scratchpad sizing only and it works fine.

I re-tested both cases: gpt-oss passes with or without the Python metadata change, but Gemma 4 still hangs with the C++ fix alone.

Gemma 4 completes only when `self.num_kv_heads = kv_cache_spec.num_kv_heads` is restored, so this metadata change seems required for Gemma 4 and does not regress gpt-oss at all.

**Comment by gausah01** [comment]
> Hmmm, just checked the hang issue still exist. Seems it it very tricky...

Thanks for checking, It seems like with the C++ scratchpad fix alone, Gemma 4 still hangs after removing the metadata change. Restoring the metadata change makes Gemma 4 complete again.

**Comment by gausah01** [comment]
Can i get another review on this please? @bigPYJ1151

File: vllm/v1/attention/backends/cpu_attn.py (hunk 5 above)
Comment (fadara01) [reply vllm/v1/attention/backends/cpu_attn.py:144]: could you help me understand why 
```
self.num_kv_heads = vllm_config.model_config.get_num_kv_heads(parallel_config)
```
and
```
self.num_kv_heads = kv_cache_spec.num_kv_heads
```

are different for gemma4?

Comment (bigPYJ1151) [reply vllm/v1/attention/backends/cpu_attn.py:144]: I see, Gemma4 has hybrid MHA configs, ```vllm_config.model_config.get_num_kv_heads(parallel_config)``` cant support it. Should use ```kv_cache_spec``` here.

**Review by bigPYJ1151 — APPROVED** [review]
Thanks @gausah01 for the investigation and fix :) 
Overall looks good, can you fix the DCO check then we can merge the PR?

**Comment by gausah01** [comment]
> Thanks @gausah01 for the investigation and fix :) Overall looks good, can you fix the DCO check then we can merge the PR?

Done, and DCO checks is passing now. Thankyou!

**Comment by gausah01** [comment]
The remaining failures seems unrelated to CPU attention path. Could someone please help rerun this failing job. cc: @fadara01 @bigPYJ1151
